Add documentation for new S3 storage class settings (#150023)
elastic/elasticsearch: docs/reference/elasticsearch/configuration-reference/s3-repository-settings.md

`storage_class`
:   Sets the S3 storage class for objects in the repository. `data_storage_class` and `metadata_storage_class` override this setting if specified, and it is generally recommended to set those values instead.

:   Accepted values are `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`.

:   Defaults to `standard` if not specified.

:   Refer to [S3 storage classes](docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.

`data_storage_class` {applies_to}`stack: ga 9.5`
:   Sets the S3 storage class for data blobs in the repository. These blobs hold the files that make up each snapshotted shard and make up most of the repository volume, but are only read when restoring a shard or accessing its contents (for example, searchable snapshots read this data).

> :   Accepted values are `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`.

Reviewer: Are you sure it wants these leading `:` characters? E.g. the `readonly` setting has multiple paragraphs of text but doesn't have these.

Author: Yeah you're right I don't believe they're [needed](https://github.com/elastic/elasticsearch/pull/150023#issuecomment-4557441163). Updated to remove them.

`storage_class`

> :   Sets the S3 storage class for objects written to the repository. Values may be `standard`, `reduced_redundancy`, `standard_ia`, `onezone_ia` and `intelligent_tiering`. Defaults to `standard`. Refer to [S3 storage classes](docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.

Reviewer: Should we suggest using the new settings instead of this one? Also do we support more values now?

Reviewer 2: \+ and mention that those settings override this one when set

Author: Updated wording in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6. LMK what you think

> `data_storage_class`

Reviewer 3: I think we want to have `applies_to` tag for both the new settings. Similar to this place: https://github.com/BrianRothermich/elasticsearch/blob/e4c1220be68f2249752d5a0f222f3257bd70a85f/docs/reference/elasticsearch/configuration-reference/s3-repository-settings.md?plain=1#L83

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6

> `data_storage_class`

Reviewer 2: ```suggestion
`data_storage_class` {applies_to}`stack: ga 9.5`
```

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6

> :   Sets the S3 storage class for data blobs written to or read from the repository. These blobs hold the files that make up each snapshotted shard and account for the bulk of the repository volume, but are only read when restoring a shard or accessing its contents. Some examples include searchable snapshots and reading from cold/frozen tiers. If this value is specified, it takes precedence over the `storage_class` value for data blob operations. Values may be `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`. Defaults to the value of `storage_class` if not specified. Refer to [S3 storage classes] (docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.

Reviewer: A few nits:

- "written to or read from" seems out-of-place here. The storage class only applies on upload but one could interpret this as meaning it may apply differently on download too. In my mind it relates more to the data at rest rather than on its way in or out as this phrase suggests.

- "searchable snapshots and reading from cold/frozen tiers" is a tautology, I think we should only say one of them (and I'd prefer "searchable snapshots").

- "takes precedence" seems a little overcomplicated for non-native speakers, and we say "Defaults to the value of `storage_class` if not specified" later which means the same thing. I would prefer just this latter presentation.

- There's a space in `] (` in the link.

Reviewer 2: \+

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6

> :   Sets the S3 storage class for data blobs written to or read from the repository. These blobs hold the files that make up each snapshotted shard and account for the bulk of the repository volume, but are only read when restoring a shard or accessing its contents. Some examples include searchable snapshots and reading from cold/frozen tiers. If this value is specified, it takes precedence over the `storage_class` value for data blob operations. Values may be `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`. Defaults to the value of `storage_class` if not specified. Refer to [S3 storage classes] (docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.

Reviewer 2: trying to avoid conflating setting values with setting precedence with `storage class`

also suggest breaking this up a little bit so the accepted values stand out a little more

```suggestion
:   Sets the S3 storage class for data blobs written to or read from the repository. These blobs hold the files that make up each snapshotted shard and make up most of the repository volume, but are only read when restoring a shard or accessing its contents. Some examples include searchable snapshots and reading from cold/frozen tiers.

    Accepted values are `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`. 

    Defaults to the value of the `storage_class` setting if not specified. 

    Refer to [S3 storage classes](docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.
```

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6

> `metadata_storage_class`

Reviewer 2: ```suggestion
`metadata_storage_class` {applies_to}`stack: ga 9.5`
```

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6

> :   Sets the S3 storage class for metadata blobs written to or read from the repository. These are generally smaller in size than data blobs but may be read more frequently for operations such as listing repository contents, taking snapshots, or otherwise manipulating the repository. If this value is specified, it takes precedence over the `storage_class` value for metadata blob operations. Values may be `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`. Defaults to the value of `storage_class` if not specified. Refer to [S3 storage classes] (docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.

Reviewer 2: ```suggestion
:   Sets the S3 storage class for metadata blobs written to or read from the repository. These are generally smaller in size than data blobs but might be read more frequently for operations such as listing repository contents, taking snapshots, or otherwise manipulating the repository.

    Accepted values are `standard`, `standard_ia`, `reduced_redundancy`, `onezone_ia`, `intelligent_tiering`, `glacier_ir`, `outposts`, and `snow`. 
    
    Defaults to the value of the `storage_class` setting if not specified. 
    
    Refer to [S3 storage classes] (docs-content://deploy-manage/tools/snapshot-and-restore/s3-repository.md#repository-s3-storage-classes) for more information.
```

Author: Done in https://github.com/elastic/elasticsearch/pull/150023/changes/f4f63bd8b3ffe29b52db884d99a8a9cef00c67c6
